src/TestResults/TestResultProvider.ts

```diff
@@ -31,26 +42,65 @@ export class TestResultProvider {
   }
 
   getResults(filePath: string): TestResult[] {
+    const maybeTemplateLiteral = (s: string) => s.indexOf('$') > -1
+
+    const findAssertionByLocation = (testBlock: ItBlock, _assertions: TestAssertionStatus[]) => {
+      return _assertions.find(
+        a =>
+          (a.line >= testBlock.start.line && a.line <= testBlock.end.line) ||
+          (a.location && a.location.line >= testBlock.start.line && a.location.line <= testBlock.end.line)
+      )
+    }
+
     if (this.resultsByFilePath[filePath]) {
       return this.resultsByFilePath[filePath]
     }
 
     const { itBlocks } = parseTest(filePath)
-    const results = this.reconciler.assertionsForTestFile(filePath) || []
+
+    const assertions = this.reconciler.assertionsForTestFile(filePath) || []
 
     const result: TestResult[] = []
     for (const test of itBlocks) {
-      const assertion =
-        // do we ever consider marking a test when their name do not match even though
-        // the linenumber matches? Especially consider line-number could be error-prone
-        // in situations lik typescript/uglify etc. without checking test names we
-        // could end up marking the wrong test simply because its line number matched.
-        // see https://github.com/jest-community/vscode-jest/issues/349
-
-        results.filter(r => r.title === test.name && r.line >= test.start.line && r.line <= test.end.line)[0] ||
-        results.filter(r => r.title === test.name && r.status !== TestReconciliationState.KnownFail)[0] ||
-        results.filter(r => r.title === test.name)[0] ||
-        ({} as any)
+      let assertion: TestAssertionStatus | undefined
+      let err: string | undefined
+
+      const match = assertions.filter(a => a.title === test.name)
+      switch (match.length) {
+        case 1:
+          assertion = match[0]
+          break
+        case 0:
+          if (maybeTemplateLiteral(test.name)) {
+            assertion = findAssertionByLocation(test, assertions)
+            if (this.verbose) {
+              // tslint:disable-next-line no-console
+              console.log(
+                `not able to match test block by name, possible due to template-iteral? matching by line number instead.`
+              )
+            }
+            if (!assertion) {
+              err = 'failed to match test result, might be caused by template-literal test name?'
+            }
+          }
+          break
+        default:
+          // multiple matches, select according to the following criteria
+          assertion = findAssertionByLocation(test, match)
+          if (!assertion) {
+            // can't find the match, it could due to sourceMap related issue, let's try our best to locate one then
+            assertion = match.find(a => a.status !== TestReconciliationState.KnownFail) || match[0]
```

Comment:
hmm... good question, I don't know the answer either... looks like it was added in this [commit](https://github.com/jest-community/vscode-jest/commit/4a5d225a31930b091e1fa0a0be85414bf66e2b82#diff-785a64be7997ebe2d7a9fdf2b7231468), maybe @ThomasRooney can shed some light?

Reply:
I think I figure this one out, I think it was assuming the "failed" test will have the line number matched already, so it is only focused on matching the non-failed ones, but of course, this assumption might not always be true since any location matching could fail due to transpile/sourcemap... 

while adding more tests to this, I discover an edge case for duplicate test names. Also didn't feel comfortable about template-literal use case, which right now will fall back to location matching, kind of dangerous as mentioned above. Added partial name match to increase correctness, when in doubt (such as multiple partial matched names found), we will just mark it as 'unknown' and log the info in console (in debug mode).

@stephtr if you get a chance, please take a look to make sure I didn't miss anything obvious, appreciate it.